apps/prairielearn/src/middlewares/authzCourseOrInstance.ts

```diff
@@ -596,7 +598,33 @@ export async function authzCourseOrInstance(req: Request, res: Response) {
       res.locals.authz_data.user_with_requested_uid_has_instructor_access_to_course_instance =
         user_with_requested_uid_has_instructor_access_to_course_instance;
     }
+  }
+
+  res.locals.authz_data.overrides = overrides;
+
+  res.locals.user = res.locals.authz_data.user;
+  res.locals.is_administrator = res.locals.authz_data.is_administrator;
+
+  res.locals.authz_data.mode = effectiveAuthzData.mode;
+  res.locals.authz_data.mode_reason = effectiveAuthzData.mode_reason;
+  res.locals.req_date = req_date;
 
+  // Recalculate this for the effective user as well.
+  if (req.params.course_instance_id) {
+    if (
+      res.locals.course_instance?.modern_publishing &&
+      res.locals.authz_data.course_instance_role === 'None'
+    ) {
+      // We use this access system instead of the legacy access system.
+      const { has_student_access, has_student_access_with_enrollment } =
+        await calculateModernCourseInstanceStudentAccess(
+          res.locals.course_instance,
+          res.locals.authz_data,
+          req_date,
+        );
+      res.locals.authz_data.has_student_access = has_student_access;
+      res.locals.authz_data.has_student_access_with_enrollment = has_student_access_with_enrollment;
+    }
```

Comment:
It sorta feels like we should be relying on `buildAuthzData` for this somehow, even in the effective user case. When constructing `effectiveAuthzData`, could we use `buildAuthzData` instead of `selectAuthzData`?

Reply:
If you want to push that complexity into buildAuthzData, I would prefer to not do it as part of this PR. `buildAuthzData` was designed to only handle the "ordinary" case, so that other pages who want to call it don't have to think about effective users.

Reply:
It seems like this is related to https://github.com/PrairieLearn/PrairieLearn/pull/13173#discussion_r2484246677. It seems like you want a refactor of the authzCourseOrInstance middleware.

Reply:
If so, we should probably do that ahead of this PR, and update locations like the enroll page to take advantage of that. https://github.com/PrairieLearn/PrairieLearn/pull/13173#discussion_r2482800107